src/main/java/loqor/ait/core/engine/block/multi/MultiBlockStructure.java

```diff
@@ -0,0 +1,209 @@
+            return new BlockOffset(this.block, this.offset.add(x, y, z));
+        }
+        public BlockOffset offset(BlockPos offset) {
+            return new BlockOffset(this.block, this.offset.add(offset));
+        }
+        public BlockOffset allow(Block... blocks) {
+            this.block.addAll(List.of(blocks));
+            return this;
+        }
+
+        @Override
+        public String toString() {
+            return "BlockOffset{" +
+                    "block=" + block +
+                    ", offset=" + offset +
+                    '}';
+        }
+
+        public static List<BlockOffset> corners(Block block, int x, int y, int z) {
+            return List.of(
+                    new BlockOffset(block, x, y, z),
+                    new BlockOffset(block, -x, y, z),
+                    new BlockOffset(block, x, -y, z),
+                    new BlockOffset(block, x, y, -z),
+                    new BlockOffset(block, -x, -y, z),
+                    new BlockOffset(block, x, -y, -z),
+                    new BlockOffset(block, -x, y, -z),
+                    new BlockOffset(block, -x, -y, -z)
+            );
+        }
+        public static List<BlockOffset> volume(Block block, int width, int height, int depth) {
+            List<BlockOffset> offsets = new ArrayList<>();
+            for (int x = 0; x < width; x++) {
+                for (int y = 0; y < height; y++) {
+                    for (int z = 0; z < depth; z++) {
+                        offsets.add(new BlockOffset(block, x, y, z));
+                    }
+                }
+            }
+            return offsets;
+        }
+        public static List<BlockOffset> square(Block block, Direction dir, int space, @Nullable Block... centre) {
+            if (centre == null) {
+                centre = new Block[]{block};
+            }
+            Block first = centre[0];
+
+            List<BlockOffset> offsets = new ArrayList<>();
+            for (int x = -space; x <= space; x++) {
+                for (int z = -space; z <= space; z++) {
+                    boolean isEdge = Math.abs(x) == space || Math.abs(z) == space;
+                    if (isEdge) {
+                        BlockOffset offset;
+                        if (dir == Direction.UP || dir == Direction.DOWN) {
+                            offset = new BlockOffset(x == 0 || z == 0 ? first : block, x, 0, z);
+                        } else if (dir == Direction.NORTH || dir == Direction.SOUTH) {
+                            offset = new BlockOffset(x == 0 || z == 0 ? first : block, x, z, 0);
+                        } else {
+                            offset = new BlockOffset(x == 0 || z == 0 ? first : block, 0, x, z);
+                        }
+                        offsets.add(offset.allow(centre));
+                    }
+                }
+            }
+
+            return offsets;
+        }
+    }
+
+    public static class AllowedBlocks extends ArrayList<Block> {
```

Comment:
If you're using `AllowedBlocks` as a sort of black/white-list then you really shouldn't use `ArrayList`. Use `Set`.
It has faster `#contains` implementation and doesn't allow for duplicate values.

Reply:
will do